### PR TITLE
Deploy script for staging

### LINKED ISSUE #8: Verify contracts on deploy
Contracts can be verified using the `hardhat verify` command. We should integrate the verification with the deployment, so we never forget to verify contracts. That's quite good for debugging also, by having the contract on etherscan we can use their interface to make simple calls.

originally from @vrde

### LINKED ISSUE #12: Improve deploy script for staging

### PR DESCRIPTION
**Pull request by vrde**
We have many contracts now, we need a more streamlined deploy and setup process. I wrote a class called `NeokingdomDAO` that takes care of this.

It has two instances:

- `NeokingdomDAOHardhat` use it together with HardHat tasks, it outputs logs in the console, and it stores things on disk.
- `NeokingdomDAOMemory` use it for testing, it keeps everything in memory.

The class is used to run "sequences", generic list of commands to do things like deploying the DAO or setting it up.

What is pretty nice is that when running a sequence using `NeokingdomDAOHardhat`, the class stores the progress in a file. In case of an exception (i.e. when there is a network issue and the transaction fails), if the sequence is run again it will start from the last checkpoint.

This is still quite experimental, but if you look at the deploy sequence you'll see how clean and simple the code is: https://github.com/NeokingdomDAO/contracts/pull/18/files#diff-52b83d9606936f7d0a369cfdb236d7bdaf3861538f4cb4c35d0843445cd71864R138-R207

Closes #12, closes #8

### CODE REVIEW

File: lib/sequences/setup.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+  ),
+  (c) =>
+    c.ResolutionManager.addResolutionType(
+      "30sNotice3mVoting",
+      66,
+      30,
+      60 * 3,
+      false
+    ),
+];
```
Comment (sirnicolaz) [inline lib/sequences/setup.ts:69]: could you replace single-letter variables with meaningful names? I am having some issues following through the code with all the x e and c

File: test/Upgrade.ts [hunk 2]
```diff
@@ -26,9 +30,9 @@ const DAY = 60 * 60 * 24;
 describe("Upgrade", () => {
   let snapshotId: string;
 
-  let token: NeokingdomToken;
-  let registry: ShareholderRegistry;
-  let resolution: ResolutionManager;
+  let NeokingdomToken: NeokingdomToken;
+  let ShareholderRegistry: ShareholderRegistry;
+  let ResolutionManager: ResolutionManager;
```
Comment (sirnicolaz) [inline test/Upgrade.ts:35]: why UpperCamelCase for variables?

File: test/Upgrade.ts [hunk 3]
```diff
@@ -43,16 +47,24 @@ describe("Upgrade", () => {
   before(async () => {
     [deployer, reserve, managingBoard, user1, user2, user3] =
       await ethers.getSigners();
-    ({ token, registry, resolution } = await deployDAO(
+    const n = await NeokingdomDAOMemory.initialize({
```
Comment (sirnicolaz) [inline test/Upgrade.ts:50]: could you rename `n` to something meaningful like `daoDeployer`, `neoKingdomDAOMemory`, `neoKingdomDAO`, `daoDeployment`, `neoKingdomDAODeployment`.... ?

File: lib/core.ts [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+          throw e;
+        }
+      }
+      if (this.config.verbose) {
+        console.log();
+      }
+      // FIXME: wait should always be a valid attribute, but it's not
+      if (tx?.wait) {
+        await tx.wait(1);
+      }
```
Comment (sirnicolaz) [inline lib/core.ts:145]: isn't this equivalent to `await tx?.wait(1)`?

File: lib/utils.ts [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+        allowFailure: false,
+        callData: populatedTx.data,
+      };
+    })
+  );
+  console.log(cd);
+
+  return c.Multicall3.aggregate3(cd);
+}
+*/
```
Comment (sirnicolaz) [inline lib/utils.ts:292]: do we still need to keep a comment here?

**Review by sirnicolaz — CHANGES_REQUESTED** [review]
Great job, awesome piece of library! There don't seem to be structural issues, only a few comments about "aesthetics". Some are directly in the code, some I write here:

- there are a bunch of functions with single-letter variable names. It's fine having `e` in the error handling code or `c` in the DEPLOY_SEQUENCE, but in the other case, it makes reading the code very tricky. Can you please use more meaningful names? I added a few comments in some specific places, but after a while, I thought it was better to write a generic request
- I would like to propose a different folder structure for the library
```
deployment/
  utils.ts
  config.ts
  sequence/
     deploy.ts
     setup.ts
  environments/
     hardhat.ts
     memory.ts
  internals
     core.ts
     types.ts   
```
- is `lib/index.ts` (empty file) necessary?

File: lib/core.ts [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+        if (force) {
+          console.error(e);
+        } else {
+          throw e;
+        }
+      }
+      if (this.config.verbose) {
+        console.log();
+      }
+      // FIXME: wait should always be a valid attribute, but it's not
```
Comment (sirnicolaz) [inline lib/core.ts:133]: why so? if `tx` is null, it makes sense that `wait` is not there, right?

Comment (vrde) [reply lib/core.ts]: The issue here is that the `step` function was returning the wrong type. I fixed it in my last commit https://github.com/NeokingdomDAO/contracts/pull/18/commits/41210e9bcc0d42046c0dd6722fb3758c37b58bfb

File: lib/utils.ts (hunk 5 above)
Comment (vrde) [reply lib/utils.ts]: removed

**Review by sirnicolaz — APPROVED** [review]

**Pull request merged**